File: examples/paper_density_simulation.R
# Small, configurable reproduction of the paper's density-band experiment.
# Usage: Rscript paper_density_simulation.R [sample_size] [replications]
#                                         [n_boot] [seed] [bandwidth_factor]
# Use 500 1000 1000 2026 1 for the first paper-scale sample size.

library(debiasedInference)

arguments <- commandArgs(trailingOnly = TRUE)
sample_size <- if (length(arguments) >= 1L) as.integer(arguments[1L]) else 500L
replications <- if (length(arguments) >= 2L) as.integer(arguments[2L]) else 10L
n_boot <- if (length(arguments) >= 3L) as.integer(arguments[3L]) else 199L
seed <- if (length(arguments) >= 4L) as.integer(arguments[4L]) else 2026L
bandwidth_factor <- if (length(arguments) >= 5L) as.double(arguments[5L]) else 1

set.seed(seed)
grid <- seq(-4, 8, length.out = 301L)
truth <- 0.6 * dnorm(grid) + 0.4 * dnorm(grid, mean = 4)
covered <- logical(replications)
widths <- numeric(replications)

for (replication in seq_len(replications)) {
  component <- runif(sample_size) >= 0.6
  sample <- rnorm(sample_size) + 4 * component
  bandwidth <- density_bandwidth(sample) * bandwidth_factor
  band <- kde_confidence_band(
    sample, grid, bandwidth = bandwidth, n_boot = n_boot,
    random_state = seed + replication
  )
  covered[replication] <- all(band$lower <= truth & truth <= band$upper)
  widths[replication] <- mean(band$upper - band$lower)
}

cat(sprintf("coverage=%.4f\n", mean(covered)))
cat(sprintf("mean_band_radius=%.6f\n", mean(widths) / 2))

File: examples/paper_density_simulation.py
"""Small, configurable reproduction of the paper's density-band experiment.

The paper uses 1,000 Monte Carlo replications and 1,000 bootstrap samples.
Defaults here are intentionally smaller for an interactive smoke run; pass
``--replications 1000 --n-boot 1000`` for the paper-scale experiment.
"""

from __future__ import annotations

import argparse

import numpy as np

from debiased_inference import density_bandwidth, kde_confidence_band


def normal_density(x: np.ndarray, mean: float) -> np.ndarray:
    return np.exp(-0.5 * (x - mean) ** 2) / np.sqrt(2.0 * np.pi)


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--sample-size", type=int, default=500)
    parser.add_argument("--replications", type=int, default=10)
    parser.add_argument("--n-boot", type=int, default=199)
    parser.add_argument("--seed", type=int, default=2026)
    parser.add_argument("--bandwidth-factor", type=float, default=1.0)
    arguments = parser.parse_args()

    generator = np.random.default_rng(arguments.seed)
    grid = np.linspace(-4.0, 8.0, 301)
    truth = 0.6 * normal_density(grid, 0.0) + 0.4 * normal_density(grid, 4.0)
    covered = 0
    widths: list[float] = []
    for replication in range(arguments.replications):
        component = generator.random(arguments.sample_size) >= 0.6
        sample = generator.normal(size=arguments.sample_size) + 4.0 * component
        bandwidth = density_bandwidth(sample) * arguments.bandwidth_factor
        band = kde_confidence_band(
            sample,
            grid,
            bandwidth=bandwidth,
            n_boot=arguments.n_boot,
            random_state=arguments.seed + replication + 1,
        )
        covered += int(np.all((band.lower <= truth) & (truth <= band.upper)))
        widths.append(float(np.mean(band.width)))

    print(f"coverage={covered / arguments.replications:.4f}")
    print(f"mean_band_width={np.mean(widths):.6f}")


if __name__ == "__main__":
    main()
